shell/packages/accounts-email-token/token_server.js

```diff
@@ -0,0 +1,163 @@
+var TOKEN_EXPIRATION_MS = 15 * 60 * 1000;
+
+var cleanupExpiredTokens = function() {
+  Meteor.users.update({}, {
+    $pull: {
+      "services.emailToken.tokens": {
+        createdAt: {$lt: new Date(Date.now() - TOKEN_EXPIRATION_MS)}
+      }
+    }
+  }, { multi: true });
+};
+
+Meteor.startup(cleanupExpiredTokens);
+// Tokens can actually last up to 2 * TOKEN_EXPIRATION_MS
+Meteor.setInterval(cleanupExpiredTokens, TOKEN_EXPIRATION_MS);
+
+Accounts._checkToken = function (user, token) {
+  var found = false;
+  user.services.emailToken.tokens.forEach(function (userToken) {
+    if((userToken.algorithm === token.algorithm) &&
+       (userToken.digest === token.digest)) {
+      found = true;
+    }
+  });
+
+  return found;
+};
+var checkToken = Accounts._checkToken;
+
+// Handler to login with a token.
+Accounts.registerLoginHandler("emailToken", function (options) {
+  if (!options.emailToken)
+    return undefined; // don't handle
+
+  options = options.emailToken;
+  check(options, {
+    email: String,
+    token: String
+  });
+
+  var user = Meteor.users.findOne({"services.emailToken.email": options.email});
+
+  if (!user) {
+    console.error("User not found:", options.email);
+    return {
+      error: new Meteor.Error(403, "User not found")
+    };
+  }
+
+  if (!user.services || !user.services.emailToken ||
+      !(user.services.emailToken.tokens)) {
+    console.error("User has no token set:", options.email);
+    return {
+      error: new Meteor.Error(403, "User has no token set")
+    };
+  }
+
+  var token = Accounts._hashToken(options.token);
+  var found = checkToken(
+    user,
+    token
+  );
+
+  if (!found) {
+    console.error("Token not found:", user.services.emailToken);
+    return {
+      error: new Meteor.Error(403, "Token not found")
+    };
+  }
+
+  Meteor.users.update({_id: user._id}, {$pull: {"services.emailToken.tokens": token}});
+  return {
+    userId: user._id
+  };
+});
+
+var Url = Npm.require("url");
+
+var ROOT_URL = Url.parse(process.env.ROOT_URL);
+var HOSTNAME = ROOT_URL.hostname;
+
+var makeTokenUrl = function (email, token) {
+  return process.env.ROOT_URL + "/_emailToken/" + encodeURIComponent(email) + "/" + encodeURIComponent(token);
+};
+
+///
+/// EMAIL VERIFICATION
+///
+var sendTokenEmail = function (email, token) {
+  var options = {
+    to:  email,
+    from: "Sandstorm " + HOSTNAME + " <no-reply@" + HOSTNAME + ">",
+    subject: "Sandstorm token for " + HOSTNAME,
+    text: "A token has requested on your behalf to login to a Sandstorm instance at " + HOSTNAME +
```

Comment:
Hmm, users not familiar with authentication stuff might be confused by the word "token". Also, let's emphasize the specific host name over "Sandstorm", which is just the software running on the host. Let's do:

    From: HOSTNAME <no-reply@HOSTNAME>
    Subject: Log in to HOSTNAME

    To log in to HOSTNAME, click on the following link:

    LINK

    Alternatively, enter the following one-time authentication code into the log-in form:

    TOKEN

    You are receiving this because someone (hopefully you) requested to log in to HOSTNAME with your email address. If you did not request to log into HOSTNAME, you may ignore this message.

    This information will expire in 15 minutes.